Feat: Progress Elements

cqliu1 opened the pull request:
Closes #23120. 
Closes #23121. 
Closes #23122.

This adds the long-awaited progress elements to Canvas. They are similar to `Reveal Image` where they take a number value as context, usually a percentage.

### Progress Shape SVG
We can turn any SVG with a defined `viewBox` and a single nested `<path>` element into a progress indicator. If you want to support labels, you just need to add a `<text>` element.
```
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 200 200">
  <path d="M 123 189 C 93 141 129 126 102 96 L 78 102 L 48 117 L 42 129 Q 30 132 21 126 L 18 114 L 27 90 L 42 72 L 48 57 L 3 6 L 57 42 L 63 33 L 60 15 L 69 27 L 69 15 L 84 27 Q 162 36 195 108 Q 174 159 123 189 Z" />
  <text x="0" y="200" text-anchor='start' dominant-baseline='text-after-edge' />
</svg>
```

### Args
<img width="241" alt="screen shot 2018-09-13 at 5 08 40 pm" src="https://user-images.githubusercontent.com/1697105/45532588-7d0a8e80-b7a9-11e8-9701-c56e76a74229.png">

#### `shape` - the shape of the progress element, i.e gauge, bar, wheel
#### `max` - upper bound of the progress element, default: `1`
#### `valueColor` - color of the progress bar
#### `barColor` - color of the background bar
#### `valueWeight` - thickness of the progress bar, default: `20`
#### `barWeight` - thickness of the background bar, default: `20`
#### `label` - set to true/false to show hide the value or set to a string to show a custom label, default: `true`
#### `font` - font settings for the label

### Shapes

#### Horizontal Bar
<img width="370" alt="screen shot 2018-09-24 at 11 51 21 am" src="https://user-images.githubusercontent.com/1697105/45972314-47a74180-bff0-11e8-9436-625092cb65a4.png">

#### Horizontal Pill
<img width="370" alt="screen shot 2018-09-24 at 11 51 14 am" src="https://user-images.githubusercontent.com/1697105/45972319-4b3ac880-bff0-11e8-8c46-70d5236da2e7.png">

### Gauge
<img width="219" alt="screen shot 2018-09-24 at 10 18 52 am" src="https://user-images.githubusercontent.com/1697105/45970095-7f12ef80-bfea-11e8-9665-e1ee3f45a168.png">

### Semicircle
<img width="188" alt="screen shot 2018-09-24 at 11 55 33 am" src="https://user-images.githubusercontent.com/1697105/45972535-da47e080-bff0-11e8-817e-f90808175bd2.png">

#### Vertical Bar
<img width="82" alt="screen shot 2018-09-24 at 11 51 07 am" src="https://user-images.githubusercontent.com/1697105/45972336-5130a980-bff0-11e8-9e88-0ab85afc2533.png">

#### Vertical Pill
<img width="76" alt="screen shot 2018-09-24 at 11 51 01 am" src="https://user-images.githubusercontent.com/1697105/45972345-55f55d80-bff0-11e8-88d0-3f5da15d8ec3.png">

#### Wheel
<img width="201" alt="screen shot 2018-09-24 at 10 17 20 am" src="https://user-images.githubusercontent.com/1697105/45970137-981ba080-bfea-11e8-91b4-2ee3d3fcca62.png">

#### (Bonus) Unicorn
<img width="252" alt="screen shot 2018-09-13 at 11 03 39 pm" src="https://user-images.githubusercontent.com/1697105/45532543-5b110c00-b7a9-11e8-869b-197b984fa3f0.png">

```
filters
| demodata
| math "sum(min(cost) / max(cost))"
| progress shape="unicorn" label={formatnumber "0%"} 
  font={font size=24 family="'Open Sans', Helvetica, Arial, sans-serif" color="#000000" align="center"} labelPosition="below" weight=5
| render 
```

## Notes
- resizing the element also resizes the label. Should the label live outside of the SVG and not be affected by scaling?
- label also shrinks as you increase the `weight`, it has to do with the label being nested in the SVG
- If you apply a thick enough `weight` to wheel, it'll look like a pie, which is a pretty nice feature. Should I include that as a separate element like `progress pie`?
<img width="238" alt="screen shot 2018-09-13 at 11 45 02 pm" src="https://user-images.githubusercontent.com/1697105/45534154-12f4e800-b7af-11e8-9507-78c2d7bcaf8b.png">

### Update (9/21/18)
You can now adjust the weight of the bars independently and produce elements like these:
<img width="165" alt="screen shot 2018-09-21 at 12 39 37 pm" src="https://user-images.githubusercontent.com/1697105/45902493-7678aa00-bd9b-11e8-81fa-c376530f996c.png">
<img width="184" alt="screen shot 2018-09-21 at 12 39 20 pm" src="https://user-images.githubusercontent.com/1697105/45902496-78db0400-bd9b-11e8-8051-134a9db44ec7.png">

### Update (9/24/18)
- removed `labelPosition` arg. We decide to have each shape define the static label position.

cqliu1 commented:
> From https://github.com/elastic/kibana-canvas/pull/1097#issuecomment-420274515
> Clicking into a vertical progress bar gives me a horizontal one. False advertising 😄
Fixed 👍 

> Also it seems odd that as soon as I delete it, I have no ability to customize the label without going into the expression language. I imagine that this could get confusing for users.
I was thinking of making the label input just a `string` input and turning the `labelPosition` form into something similar to `axisConfig` with the toggle to show/hide and the position select input

> Probably out of scope for this PR, but I can't seem to resize to stop cutting off the element no matter how big or small I make the container.
I fixed the calculations for the SVG dimensions and resizing should adjust in respect to `weight` and `label` dimensions

> And last but not least, a simple question. If there is no max limit, what's the math behind showing progress when the value isn't a percentage?
The percentage is `context` / `max`. Max is set to 1 by default, so it works only for values. If you pass in a value > 1, you'll want to increase the max.

cqliu1 commented:
@rashidkpc I made all the changes according your feedback from https://github.com/elastic/kibana-canvas/pull/1097#pullrequestreview-154218445. I also made the progress shape definition simpler, so it's a little more pluggable.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/2569/)

rashidkpc commented:
@alexfrancoeur do you have time to give this another once over? I'm going to check it out today.

alexfrancoeur commented:
yeah, will try within the next few hours

alexfrancoeur commented:
Feedback below (in no prioritized order). 

First I noticed that the element selection window seems a bit condensed. Is this related to this PR or is it something else? @rashidkpc 

<img width="1677" alt="screen shot 2018-09-18 at 5 29 34 pm" src="https://user-images.githubusercontent.com/25181823/45717611-85c3e180-bb68-11e8-8f0b-2b6415a35734.png">

Preview of a dataset also seems to be weird.

<img width="1675" alt="screen shot 2018-09-18 at 5 32 49 pm" src="https://user-images.githubusercontent.com/25181823/45717771-e6531e80-bb68-11e8-9329-c7444b265788.png">

I can open separate bugs for those tomorrow if preferred ^^

@cqliu1 @rashidkpc for metrics and progress elements, wouldn't it make more sense to have a column that's a percentage in the demo data? This way we don't need to do this every time. It's such an odd experience because a custom script like that is the default option but not available directly in the UI.

![sep-18-2018 17-37-58](https://user-images.githubusercontent.com/25181823/45718009-a6406b80-bb69-11e8-8081-9610a21fe6ad.gif)

The same could be said about this. As soon as I delete it, it's gone. Fingers crossed for https://github.com/elastic/kibana/issues/23111 🤞 🤞 🤞 🤞 

<img width="352" alt="screen shot 2018-09-18 at 5 39 40 pm" src="https://user-images.githubusercontent.com/25181823/45718070-dc7deb00-bb69-11e8-8504-09b54af0df73.png">

<img width="315" alt="screen shot 2018-09-18 at 5 40 57 pm" src="https://user-images.githubusercontent.com/25181823/45718135-0505e500-bb6a-11e8-844b-5adb675956fe.png">

TSVB has styling options for the width of the background and progress lines. If it's not too hard to add, that would be some pretty great functionality if we split the weight into background and progress weight.

<img width="1495" alt="screen shot 2018-09-18 at 5 42 16 pm" src="https://user-images.githubusercontent.com/25181823/45718189-35e61a00-bb6a-11e8-9e76-0515344a4cc5.png">

This happens to me very frequently in Canvas, but I changing elements does not update the appropriate UI. Seems to be occurring with progress elements as well. I believe we're already tracking here" https://github.com/elastic/kibana/issues/23175

![sep-18-2018 17-44-39](https://user-images.githubusercontent.com/25181823/45718318-99704780-bb6a-11e8-81a9-a1b27e3b29bd.gif)

Changing the label size seems to off center the element. Should we keep it centered? I'm on the fence as to the best approach here

![sep-18-2018 17-47-20](https://user-images.githubusercontent.com/25181823/45718417-ece29580-bb6a-11e8-9c3e-28d5a83fcb17.gif)

Other than those comments, LGTM. Can open any other bugs as they come along.

cqliu1 commented:
> First I noticed that the element selection window seems a bit condensed. Is this related to this PR or is it something else?

It's not a result of the changes in this PR. EUI applies a `max-width` of 768px to all modals that was added in a recent update. We can override the style if we want the modal to be wider in a separate PR.

rashidkpc commented:
@alexfrancoeur Please open a new ticket for the modal stuff. Especially the overflow of the preview.

rashidkpc commented:
@alexfrancoeur agree on the percent column in demodata
@cqliu1 can you add a column to demodata that contains a number between 0 and 1? Then you can just use mean(whatever) as the value.

alexfrancoeur commented:
Opened https://github.com/elastic/kibana/issues/23324 and https://github.com/elastic/kibana/issues/23325 for the modal window issues. For percent column, it'd be great to default to this with metric, image reveal and this progress elements. Not sure if I'm missing any others.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/2983/)

cqliu1 commented:
@alexfrancoeur @rashidkpc I added a percentage column to `demodata`, and I updated the expressions for image reveal and the progress element. I left the metric element the way it was because it doesn't expect a percentage value.

cqliu1 commented:
> Something I did notice, and I can't tell if if it's just me, but I don't think the Gauge label is centered vertically. Can you double check?

 The label is at the center of the shape which is different from the center of the circle. I can shift the position down a little.

rashidkpc commented on x-pack/plugins/canvas/canvas_plugin_src/renderers/progress/progress.js:
```diff
@@ -0,0 +1,156 @@
+    const shapeDef = shapes[shape];
+
+    if (shapeDef) {
+      const { path, getViewBox, width, height } = shapeDef;
```
`shapeDef` is effectively a new API. Can you explain the parameters? Do we actually need getViewBox or can we simplify this?

cqliu1 replied:
I added the `getViewBox` function to allow the user to define how the weight adjusts the shapes viewBox to keep full shape in view as the weight increase. I realized that I can apply a general calculation to all shapes to account for the stroke width, similar to how it's handled in the `shape` element.

rashidkpc commented:
I'd say the API is getting better here, but I wonder if we can get this down to a "raw" svg. For example, I quickly drew this in BoxySVG:

```
<?xml version="1.0" encoding="utf-8"?>
<svg viewBox="110.608 85.73 202.998 267.3" width="202.998" height="267.3" xmlns="http://www.w3.org/2000/svg">
  <path d="M 116.15 85.73 C 116.15 119.45 116.249 350.762 115.597 350.111 C 113.574 348.087 107.305 231.077 112.831 205.752 C 122.816 159.989 173.112 173.36 166.482 205.752 C 160.033 237.261 171.535 379.806 200.221 346.792 C 205.084 341.195 222.271 225.359 221.239 226.217 C 172.741 266.557 285.935 383.242 313.606 345.686" style="stroke: rgb(0, 0, 0); fill: none;"/>
</svg>
```

Could we just define the API as a single SVG element with a height and width, containing a single path? Then we could just duplicate the path and recolor it?

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3184/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3278/)

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/3284/)

cqliu1 commented:
test this

rashidkpc commented:
I'm good with this, but I'd like to chat with @alexfrancoeur before we merge it.

rashidkpc commented:
@cqliu1 go ahead and merge this, we can always make small tweaks later

cqliu1 commented:
6.x: f77be52
